### PR TITLE
Add detailed scanner trace output and notifications

### PR DESCRIPTION
**Pull request by klauspost**
## Description

Add `s3:Scanner:ManyVersions` and `s3:Scanner:BigPrefix` bucket notifications when seeing big prefixes or objects with many versions.

Furthermore details are sent with trace commands.

Requires: https://github.com/minio/madmin-go/pull/176

## How to test this PR?

I will add an mc version as soon as madmin is merged.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/data-scanner.go [hunk 1]
```diff
@@ -59,6 +60,9 @@ const (
 	healDeleteDangling    = true
 	healFolderIncludeProb = 32  // Include a clean folder one in n cycles.
 	healObjectSelectProb  = 512 // Overall probability of a file being scanned; one in n.
+
+	dataScannerExcessiveVersionsThreshold = 1000  // Issue a warning when a single object has more versions than this
```
Comment (harshavardhana) [inline cmd/data-scanner.go:64]: Shouldn't this be a problem even at 100 ? I can't for see anyone needing more than 100 versions

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-erasure.sh | :x: [more...](#eff282f4-b820-41af-822e-7256120f104a) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#33f6641d-a3c0-44dc-8f9d-54b5430234e6) |
| mint-pools.sh | :x: [more...](#5ff66fd6-21ec-42ce-b0dd-7508b97f917e) |
<a name="33f6641d-a3c0-44dc-8f9d-54b5430234e6"></a>
### 16668-1c7c008/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.8:31290
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp a46e120250e0:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 9 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 59 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 16 seconds
(8/14) Running minio-go tests ... done in 47 seconds
(9/14) Running minio-java tests ... FAILED in 2 minutes and 1 seconds
{
  "name": "minio-java",
  "function": "listObjects()",
  "args": "[bucket, prefix, recursive, 1050 objects]",
  "duration": 113428,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = BucketNotEmpty, message = The bucket you tried to delete is not empty, bucketName = minio-java-test-3rvcdg8, objectName = null, resource = /minio-java-test-3rvcdg8, requestId = 17458F78BACCB17A, hostId = b81cba30-2c59-49f7-9b2f-863a0f3703db)\nrequest={method=DELETE, url=http://15.15.15.8:31290/minio-java-test-3rvcdg8, headers=Host: 15.15.15.8:31290\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.4.6\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20230220T144304Z\nAuthorization: ██\n}\nresponse={code=409, headers=Accept-Ranges: bytes\nContent-Length: 328\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nStrict-Transport-Security: max-age=31536000; includeSubDomains\nVary: Origin\nVary: Accept-Encoding\nX-Amz-Id-2: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nX-Amz-Request-Id: 17458F78BACCB17A\nX-Content-Type-Options: nosniff\nX-Xss-Protection: 1; mode=block\nDate: Mon, 20 Feb 2023 14:43:04 GMT\n}\n >>> [io.minio.S3Base$1.onResponse(S3Base.java:690), okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519), java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128), java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628), java.base/java.lang.Thread.run(Thread.java:829)]"
}
(9/14) Running minio-js tests ... done in 50 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 39 seconds
(11/14) Running s3cmd tests ... done in 15 seconds
(12/14) Running s3select tests ... done in 3 seconds
(13/14) Running versioning tests ... done in 3 minutes and 11 seconds

Executed 13 out of 14 tests successfully.
```
<a name="5ff66fd6-21ec-42ce-b0dd-7508b97f917e"></a>
### 16668-1c7c008/mint-pools.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.2:32405
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp d3fc59377d32:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 9 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 1 minutes and 0 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... done in 16 seconds
(8/14) Running minio-go tests ... FAILED in 1 minutes and 55 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-pmnuc25xzdn31irk",
      "Object": "srcObject",
      "Encryption": {},
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "source": {
      "Bucket": "minio-go-test-pmnuc25xzdn31irk",
      "Object": "srcObject",
      "VersionID": "",
      "MatchETag": "",
      "NoMatchETag": "",
      "MatchModifiedSince": "0001-01-01T00:00:00Z",
      "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
      "MatchRange": false,
      "Start": 0,
      "End": 0,
      "Encryption": null
    }
  },
  "duration": 62023,
  "error": "read tcp 172.17.0.3:54160->15.15.15.2:32405: read: connection reset by peer",
  "function": "CopyObject(destination, source)",
  "message": "ReadAll failed",
  "name": "minio-go: testEncryptedSSES3ToSSES3CopyObject",
  "status": "FAIL"
}
(8/14) Running minio-java tests ... done in 36 seconds
(9/14) Running minio-js tests ... done in 52 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 51 seconds
(11/14) Running s3cmd tests ... done in 16 seconds
(12/14) Running s3select tests ... done in 5 seconds
(13/14) Running versioning tests ... done in 3 minutes and 12 seconds

Executed 13 out of 14 tests successfully.
```
<a name="eff282f4-b820-41af-822e-7256120f104a"></a>
### 16668-1c7c008/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.1:30793
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp ebacb1b69fd9:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 57 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... done in 13 seconds
(8/14) Running minio-go tests ... done in 33 seconds
(9/14) Running minio-java tests ... done in 17 seconds
(10/14) Running minio-js tests ... FAILED in 1 minutes and 13 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-2e817de9-5e2d-41ae-9a0a-07a0cb2342c8, events: s3:ObjectCreated:*",
  "duration": 63164,
  "status": "FAIL",
  "error": "Error: read ECONNRESET at TCP.onStreamRead (internal/stream_base_commons.js:209:20)"
}
(10/14) Running minio-py tests ... done in 1 minutes and 8 seconds
(11/14) Running s3cmd tests ... done in 15 seconds
(12/14) Running s3select tests ... done in 2 seconds
(13/14) Running versioning tests ... done in 3 minutes and 8 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/data-scanner.go (hunk 1 above)
Comment (klauspost) [reply cmd/data-scanner.go:64]: 100 will not significantly impact it:





Green bars.

Comment (klauspost) [reply cmd/data-scanner.go:64]: Might as well make it less likely to be noisy when it doesn't really affect us much.

Comment (harshavardhana) [reply cmd/data-scanner.go:64]: These can't be tested in SILO right?

A prefix with many objects each object having 100versions each the cost will eventually add up.

There is no requirement for any application to have anywhere close to 100 versions. Atleast I can't think of any.

Comment (klauspost) [reply cmd/data-scanner.go:64]: @harshavardhana It is separate concerns. Many files/folder doesn't affect individual version access. 

Each scales separately, so it isn't a concern if they are combined.

A prefix with 1M objects will be just as slow whether there are 1 or 1000 versions of each. So it makes sense to test them individually and report if either is "unreasonable". Keeping this as non-noisy as possible gives it more value.

We shouldn't dictate usage, except when it affects performance.

Comment (harshavardhana) [reply cmd/data-scanner.go:64]: Okay I think we can live with 1000 for now.

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#3eab76f9-1b21-4b1d-b6fd-af9268409ad5) |
<a name="3eab76f9-1b21-4b1d-b6fd-af9268409ad5"></a>
### 16668-1196e55/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.1:30128
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp b9d78a887568:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... 
Aborting Mint...
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**